In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
import os

In [3]:
raw_data = pd.read_csv(os.path.join("Resources", "Data", "growth_designation.csv"))
raw_data

Region  FIPS  ColonyCount  ColonyGrowth  GrowthOutcome  TotalProduction  \
0         3     1        15000         -6.25            0.0           960000   
1         3     1        14000         -6.67            0.0           924000   
2         3     1        16000         14.29            1.0          1136000   
3         3     1        17000          6.25            1.0          1156000   
4         3     1        16000         -5.88            0.0          1248000   
..      ...   ...          ...           ...            ...              ...   
889       4    56        50000         42.86            1.0          2550000   
890       4    56        47000         -6.00            0.0          3102000   
891       4    56        38000        -19.15            0.0          2318000   
892       4    56        38000          0.00            0.0          2926000   
893       4    56        40000          5.26            1.0          2720000   

     YieldPerColony  PricePerLB  ProductionValue  Stocks  YearRecorded  \
0                64        0.87           835000   96000          1996   
1                66        0.81           748000   92000          1997   
2                71        0.72           818000  159000          1998   
3                68        0.56           647000  185000          1999   
4                78        0.59           736000  187000          2000   
..              ...         ...              ...     ...           ...   
889              51        1.87          4769000  459000          2012   
890              66        2.11          6545000  558000          2013   
891              61        2.08          4821000  255000          2014   
892              77        1.90          5559000  146000          2015   
893              68        1.78          4842000  190000          2016   

     ClothianidinLB  ImidaclopridLB  ThiamethoxamLB  AcetamipridLB  \
0              0.00          819.24            0.00            0.0   
1              0.00        14781.55            0.00            0.0   
2              0.00         4048.35            0.00            0.0   
3              0.00         2758.42            0.00            0.0   
4              0.00         3305.17            0.00            0.0   
..              ...             ...             ...            ...   
889         2590.21          610.68          933.22            0.0   
890         1853.87          343.70         1160.07            0.0   
891         2782.23          569.23         1248.04            0.0   
892            0.00          252.21           41.89            0.0   
893            0.00           60.63           20.72            0.0   

     ThiaclopridLB  CombinedNeonicLB  
0              0.0            819.24  
1              0.0          14781.55  
2              0.0           4048.35  
3              0.0           2758.42  
4              0.0           3305.17  
..             ...               ...  
889            0.0           4134.11  
890            0.0           3357.64  
891            0.0           4599.50  
892            0.0            294.10  
893            0.0             81.35  

[894 rows x 17 columns]

In [4]:
classifier = LogisticRegression()
classifier

LogisticRegression()

In [13]:
print(f"ColonyCount: {raw_data['ColonyCount'].min()}, {raw_data['ColonyCount'].max()}")
print(f"TotalProduction: {raw_data['TotalProduction'].min()}, {raw_data['TotalProduction'].max()}")
print(f"YieldPerColony: {raw_data['YieldPerColony'].min()}, {raw_data['YieldPerColony'].max()}")
print(f"PricePerLB: {raw_data['PricePerLB'].min()}, {raw_data['PricePerLB'].max()}")
print(f"ProductionValue: {raw_data['ProductionValue'].min()}, {raw_data['ProductionValue'].max()}")
print(f"Stocks: {raw_data['Stocks'].min()}, {raw_data['Stocks'].max()}")
print(f"ClothianidinLB: {raw_data['ClothianidinLB'].min()}, {raw_data['ClothianidinLB'].max()}")
print(f"ImidaclopridLB: {raw_data['ImidaclopridLB'].min()}, {raw_data['ImidaclopridLB'].max()}")
print(f"ThiamethoxamLB: {raw_data['ThiamethoxamLB'].min()}, {raw_data['ThiamethoxamLB'].max()}")
print(f"AcetamipridLB: {raw_data['AcetamipridLB'].min()}, {raw_data['AcetamipridLB'].max()}")
print(f"ThiaclopridLB: {raw_data['ThiaclopridLB'].min()}, {raw_data['ThiaclopridLB'].max()}")
print(f"CombinedNeonicLB: {raw_data['CombinedNeonicLB'].min()}, {raw_data['CombinedNeonicLB'].max()}")

ColonyCount: 2000, 510000
TotalProduction: 84000, 46410000
YieldPerColony: 19, 128
PricePerLB: 0.47, 7.09
ProductionValue: 106000, 83859000
Stocks: 8000, 13800000
ClothianidinLB: 0.0, 613984.75
ImidaclopridLB: 0.0, 331948.48
ThiamethoxamLB: 0.0, 142935.83
AcetamipridLB: 0.0, 80425.29
ThiaclopridLB: 0.0, 9420.79
CombinedNeonicLB: 0.0, 888488.49


<h3 style="color:green">Model 1</h3>
<strong>Features: </strong>All columns

In [6]:
X1 = raw_data.drop(columns=['GrowthOutcome', 'ColonyGrowth', 'YearRecorded','FIPS','Region'])
y1 = raw_data["GrowthOutcome"]
print(X1.shape, y1.shape)
X1.head()

(894, 12) (894,)


ColonyCount  TotalProduction  YieldPerColony  PricePerLB  ProductionValue  \
0        15000           960000              64        0.87           835000   
1        14000           924000              66        0.81           748000   
2        16000          1136000              71        0.72           818000   
3        17000          1156000              68        0.56           647000   
4        16000          1248000              78        0.59           736000   

   Stocks  ClothianidinLB  ImidaclopridLB  ThiamethoxamLB  AcetamipridLB  \
0   96000             0.0          819.24             0.0            0.0   
1   92000             0.0        14781.55             0.0            0.0   
2  159000             0.0         4048.35             0.0            0.0   
3  185000             0.0         2758.42             0.0            0.0   
4  187000             0.0         3305.17             0.0            0.0   

   ThiaclopridLB  CombinedNeonicLB  
0            0.0            819.24  
1            0.0          14781.55  
2            0.0           4048.35  
3            0.0           2758.42  
4            0.0           3305.17

In [6]:
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, random_state=1)

In [7]:
classifier.fit(X1_train, y1_train)

C:\Users\Lauren\anaconda3\envs\PythonAdv\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

In [8]:
print(f"Training Data Score: {classifier.score(X1_train, y1_train)}")
print(f"Testing Data Score: {classifier.score(X1_test, y1_test)}")

Training Data Score: 0.6044776119402985
Testing Data Score: 0.6071428571428571


In [9]:
predictions1 = classifier.predict(X1_test)
print(f"First 10 Predictions:   {predictions1[:10]}")
print(f"First 10 Actual labels: {y1_test[:10].tolist()}")

First 10 Predictions:   [0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
First 10 Actual labels: [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0]


In [10]:
pd.DataFrame({"Prediction": predictions1, "Actual": y1_test}).reset_index(drop=True)

Prediction  Actual
0           0.0     0.0
1           0.0     0.0
2           0.0     0.0
3           0.0     1.0
4           0.0     0.0
..          ...     ...
219         0.0     0.0
220         1.0     0.0
221         0.0     0.0
222         1.0     1.0
223         1.0     0.0

[224 rows x 2 columns]

<h3 style="color:green">Model 2</h3>
<strong>Features: </strong>ProductionValue, TotalProduction, ImidaclopridLB, Stocks, PricePerLB
<br>(top five from random forest model)

In [11]:
X2 = raw_data[['ProductionValue', 'TotalProduction', 'ImidaclopridLB','Stocks','PricePerLB']]
y2 = raw_data["GrowthOutcome"]
# X2.head()
# y2.head()
print(X2.shape, y2.shape)

(894, 5) (894,)


In [12]:
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, random_state=1)

In [13]:
classifier.fit(X2_train, y2_train)
print(f"Training Data Score: {classifier.score(X2_train, y2_train)}")
print(f"Testing Data Score: {classifier.score(X2_test, y2_test)}")

Training Data Score: 0.5970149253731343
Testing Data Score: 0.5580357142857143


In [14]:
predictions2 = classifier.predict(X2_test)
print(f"First 10 Predictions:   {predictions2[:10]}")
print(f"First 10 Actual labels: {y2_test[:10].tolist()}")

First 10 Predictions:   [0. 0. 0. 0. 0. 1. 0. 1. 0. 0.]
First 10 Actual labels: [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0]


In [15]:
pd.DataFrame({"Prediction": predictions2, "Actual": y2_test}).reset_index(drop=True)

Prediction  Actual
0           0.0     0.0
1           0.0     0.0
2           0.0     0.0
3           0.0     1.0
4           0.0     0.0
..          ...     ...
219         0.0     0.0
220         1.0     0.0
221         0.0     0.0
222         0.0     1.0
223         1.0     0.0

[224 rows x 2 columns]

<h3 style="color:green">Model 3</h3>
<strong>Features: </strong> ClothianidinLB, ImidaclopridLB, ThiamethoxamLB, AcetamipridLB, ThiaclopridLB
<br>(each neonic pesticide)

In [16]:
X3 = raw_data[['ClothianidinLB', 'ImidaclopridLB', 'ThiamethoxamLB','AcetamipridLB','ThiaclopridLB']]
y3 = raw_data["GrowthOutcome"]
# X3.head()
# y3.head()
print(X3.shape, y3.shape)

(894, 5) (894,)


In [17]:
X3_train, X3_test, y3_train, y3_test = train_test_split(X3, y3, random_state=1)
classifier.fit(X3_train, y3_train)
print(f"Training Data Score: {classifier.score(X3_train, y3_train)}")
print(f"Testing Data Score: {classifier.score(X3_test, y3_test)}")

Training Data Score: 0.6447761194029851
Testing Data Score: 0.6517857142857143


In [18]:
predictions3 = classifier.predict(X3_test)
print(f"First 10 Predictions:   {predictions3[:10]}")
print(f"First 10 Actual labels: {y3_test[:10].tolist()}")

First 10 Predictions:   [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
First 10 Actual labels: [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0]


In [19]:
pd.DataFrame({"Prediction": predictions3, "Actual": y3_test}).reset_index(drop=True)

Prediction  Actual
0           0.0     0.0
1           0.0     0.0
2           0.0     0.0
3           0.0     1.0
4           0.0     0.0
..          ...     ...
219         0.0     0.0
220         0.0     0.0
221         0.0     0.0
222         0.0     1.0
223         0.0     0.0

[224 rows x 2 columns]

<h3 style="color:green">Model 4</h3>
<strong>Features: </strong> CombinedNeonicLB

In [22]:
X4 = raw_data[["CombinedNeonicLB"]]
y4 = raw_data["GrowthOutcome"]
# X4.head()
# y4.head()
print(X4.shape, y4.shape)

(894, 1) (894,)


In [23]:
X4_train, X4_test, y4_train, y4_test = train_test_split(X4, y4, random_state=1)
classifier.fit(X4_train, y4_train)
print(f"Training Data Score: {classifier.score(X4_train, y4_train)}")
print(f"Testing Data Score: {classifier.score(X4_test, y4_test)}")

Training Data Score: 0.6417910447761194
Testing Data Score: 0.6651785714285714


In [24]:
predictions4 = classifier.predict(X4_test)
print(f"First 10 Predictions:   {predictions4[:10]}")
print(f"First 10 Actual labels: {y4_test[:10].tolist()}")

First 10 Predictions:   [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
First 10 Actual labels: [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0]


In [25]:
pd.DataFrame({"Prediction": predictions4, "Actual": y4_test}).reset_index(drop=True)

Prediction  Actual
0           0.0     0.0
1           0.0     0.0
2           0.0     0.0
3           0.0     1.0
4           0.0     0.0
..          ...     ...
219         0.0     0.0
220         0.0     0.0
221         0.0     0.0
222         0.0     1.0
223         0.0     0.0

[224 rows x 2 columns]

<h3 style="color:green">Model 5</h3>
<strong>Features: </strong> ColonyCount, TotalProduction, YieldPerColony, PricePerLB, ProductionValue, Stocks
<br> (excludes all neonic pesticide data)

In [27]:
X5 = raw_data[['ColonyCount', 'TotalProduction', 'YieldPerColony', 'PricePerLB', 'ProductionValue', 'Stocks']]
y5 = raw_data["GrowthOutcome"]
# X5.head()
# y5.head()
print(X5.shape, y5.shape)

(894, 6) (894,)


In [28]:
X5_train, X5_test, y5_train, y5_test = train_test_split(X5, y5, random_state=1)
classifier.fit(X5_train, y5_train)
print(f"Training Data Score: {classifier.score(X5_train, y5_train)}")
print(f"Testing Data Score: {classifier.score(X5_test, y5_test)}")

Training Data Score: 0.5850746268656717
Testing Data Score: 0.5758928571428571


In [29]:
predictions5 = classifier.predict(X5_test)
print(f"First 10 Predictions:   {predictions5[:10]}")
print(f"First 10 Actual labels: {y5_test[:10].tolist()}")

First 10 Predictions:   [0. 0. 0. 0. 0. 1. 0. 0. 1. 0.]
First 10 Actual labels: [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0]


In [30]:
pd.DataFrame({"Prediction": predictions5, "Actual": y5_test}).reset_index(drop=True)

Prediction  Actual
0           0.0     0.0
1           0.0     0.0
2           0.0     0.0
3           0.0     1.0
4           0.0     0.0
..          ...     ...
219         0.0     0.0
220         1.0     0.0
221         0.0     0.0
222         1.0     1.0
223         1.0     0.0

[224 rows x 2 columns]

<h2 style= "color:green">Model Review</h2>
<strong>Model 1:</strong><br>
<ul> Training Data Score: 0.6044776119402985</ul>
<ul>Testing Data Score: 0.6071428571428571</ul>
<strong>Model 2</strong>
<ul> Training Data Score: 0.5970149253731343</ul>
<ul>Testing Data Score: 0.5580357142857143</ul>
<strong>Model 3</strong>
<ul> Training Data Score: 0.6447761194029851</ul>
<ul>Testing Data Score: 0.6517857142857143</ul>
<strong>Model 4</strong>
<ul> Training Data Score: 0.6417910447761194</ul>
<ul>Testing Data Score: 0.6651785714285714</ul>
<strong>Model 5</strong>
<ul> Training Data Score: 0.5850746268656717</ul>
<ul>Testing Data Score: 0.5758928571428571</ul>